File: Python/Samsung_B_Type_Prep/035_036_Coffee_and_Bakery/035_cpp/answer.cpp
/*
* @file: [H2534][Pro] 커피점 & 제과점
   
* @brief: 모범 답안
   
* @copyright: All rights reserved (c) 2025 Samsung Electronics, Inc.
*/
  
  
  
  
  
  
#include <queue>
#include <vector>
#include <cstring>
  
  
using namespace std;
  
  
constexpr int MAX_N = 10000;
  
  
typedef unsigned long long ull;
  
  
vector<int> Graph[MAX_N];
  
  
void init(int N, int K, int sBuilding[], int eBuilding[], int mDistance[]) {
    for (int i = 0; i < N; ++i) {
        Graph[i].clear();
    }
    for (int i = 0; i < K; ++i) {
        Graph[sBuilding[i]].emplace_back(eBuilding[i] << 16 | mDistance[i]);
        Graph[eBuilding[i]].emplace_back(sBuilding[i] << 16 | mDistance[i]);
    }
}
  
  
void add(int sBuilding, int eBuilding, int mDistance) {
    Graph[sBuilding].emplace_back(eBuilding << 16 | mDistance);
    Graph[eBuilding].emplace_back(sBuilding << 16 | mDistance);
}
  
  
int calculate(int M, int mCoffee[], int P, int mBakery[], int R) {
    int cDist[MAX_N], bDist[MAX_N];
    memset(cDist, 0x7f, sizeof(cDist));
    memset(bDist, 0x7f, sizeof(bDist));
  
  
    priority_queue<ull, vector<ull>, greater<ull>> pq;
  
  
    for (int i = 0; i < M; ++i) {
        cDist[mCoffee[i]] = 0;
        pq.emplace(1 << 16 | mCoffee[i]);
    }
  
  
    for (int i = 0; i < P; ++i) {
        bDist[mBakery[i]] = 0;
        pq.emplace(mBakery[i]);
    }
  
  
    int ret = R * 2 + 1;
    while (!pq.empty()) {
        ull curr = pq.top();
        pq.pop();
  
  
        ull currCost = curr >> 32;
        ull isCoffee = (curr >> 16) & 0xffff;
        ull u = curr & 0xffff;
  
  
        if (currCost >= ret || currCost > R) break;
  
  
        if (isCoffee) {
            if (cDist[u] < currCost) continue;
            if (currCost > 0 && currCost <= R && bDist[u] > 0 && bDist[u] <= R && currCost + bDist[u] < ret) {
                ret = currCost + bDist[u];
            }
  
  
            for (int val : Graph[u]) {
                ull v = val >> 16;
                ull nextCost = currCost + val & 0xffff;
  
  
                if (nextCost >= ret) continue;
                if (cDist[v] > nextCost) {
                    cDist[v] = nextCost;
                    pq.emplace(nextCost << 32 | 1 << 16 | v);
                }
            }
        } else {
            if (bDist[u] < currCost) continue;
            if (currCost > 0 && currCost <= R && cDist[u] > 0 && cDist[u] <= R && currCost + cDist[u] < ret) {
                ret = currCost + cDist[u];
            }
  
  
            for (int val : Graph[u]) {
                ull v = val >> 16;
                ull nextCost = currCost + val & 0xffff;
  
  
                if (nextCost >= ret) continue;
                if (bDist[v] > nextCost) {
                    bDist[v] = nextCost;
                    pq.emplace(nextCost << 32 | v);
                }
            }
        }
    }
  
  
    return (ret <= 2 * R) ? ret : -1;
}

File: Python/Samsung_B_Type_Prep/035_036_Coffee_and_Bakery/036_python/answer.py
import heapq

def init(N, K, sBuilding, eBuilding, mDistance):
	global graph, gN
	graph = [[] for _ in range(N+1)]
	gN = N
	for i in range(K) :
		graph[sBuilding[i]].append((eBuilding[i], mDistance[i]))
		graph[eBuilding[i]].append((sBuilding[i], mDistance[i]))
	return


def add(sBuilding, eBuilding, mDistance):
	global graph
	graph[sBuilding].append((eBuilding, mDistance))
	graph[eBuilding].append((sBuilding, mDistance))
	return


def calculate(M, mCoffee, P, mBakery, R):
	global graph, gN
	ans = 987654321
	hq = []
	visit = [[-1] * (gN+1) for _ in range(2) ]
	for i in range(M):
		heapq.heappush(hq, (0, mCoffee[i], 0))
	for i in range(P):
		heapq.heappush(hq, (0, mBakery[i], 1))

	while hq :
		dist, tId, CB = heapq.heappop(hq)
		if visit[CB][tId] != -1:
			continue
		visit[CB][tId] = dist

		if visit[0][tId] > 0 and visit[1][tId] > 0 :
			if ans > visit[0][tId] + visit[1][tId]:
				ans = visit[0][tId] + visit[1][tId]
			else :
				continue

		for nxt in graph[tId] :
			if dist + nxt[1] > R or dist + nxt[1] > ans or visit[CB][nxt[0]] != -1: 
				continue
			heapq.heappush(hq, (dist + nxt[1], nxt[0], CB))

	if ans == 987654321 : return -1
	return ans